graph/nullable.md

```diff
@@ -0,0 +1,312 @@
+In the case that no value is provided when the entity is created, this means that the service will create one; this value can be specified with the `DefaultValue` attribute, but if the value is contextual and determine at request time, then the property can both be non-nullable *and* have no `DefaultValue` specified. 
+Below are some examples of nullable and non-nullable properties. 
+
+## CSDL
+
+```xml
+<EntitySet Name="servicePrincipals" Type="self.servicePrincipal" />
+...
+<EntityType Name="servicePrincipal">
+  <Key>
+    <PropertyRef Name="id" />
+  </Key>
+  <Property Name="id" Type="Edm.String" Nullable="false" />
+
+  <Property Name="appId" Type="Edm.String" Nullable="false" /> <!--required for creation-->
+  <Property Name="displayName" Type="Edm.String" Nullable="false" />
+  <Property Name="foo" Type="Edm.String" Nullable="true" DefaultValue="testval" />
+  <Property Name="bar" Type="Edm.String" Nullable="false" DefaultValue="differentvalue" />
+  ...
+</EntityType>
+```
+
+## HTTP Requests
+
+### {1} Create a servicePrincipal with no properties
+
+```HTTP
+POST /servicePrincipals
+
+400 Bad Request
+{
+  "error": {
+    "code": "BadRequest",
+    "message": "The 'appId' property is required to create a servicePrincipal."
+  }
+}
+```
+
+### {2} Create a servicePrincipal without a display name
+
+```HTTP
+POST /servicePrincipals
+{
+  "appId": "00000000-0000-0000-0000-000000000001"
+}
+
+201 Created
+{
+  "appId": "00000000-0000-0000-0000-000000000001",
+  "displayName": "some application name",
+  "foo": "testval",
+  "bar": "differentvalue",
+  ...
+}
+```
+Notes:
+1. `displayName` was given a value by the service even though no value was provided by the client
+2. `foo` has the default value as specified by its `DefaultValue` attribute in the CSDL
+3. `bar` has the default value as specified by its `DefaultValue` attribute in the CSDL
+
+### {3} Update the display name of a service principal to null
+
+```HTTP
+PATCH /servicePrincipals/00000000-0000-0000-0000-000000000001
+{
+  "displayName": null
+}
+
+400 Bad Request
+{
+  "error": {
+    "code": "BadRequest",
+    "message": "null is not a valid value for the property 'displayName'; 'displayName' is not a nullable property."
+  }
+}
+```
+Notes:
+1. `displayName` cannot be set to `null` because it has be marked with `Nullable="false"` in the CSDL.
+
+### {4} Update the display name of a service principal
+
+```HTTP
+PATCH /servicePrincipals/00000000-0000-0000-0000-000000000001
+{
+  "displayName": "a non-generated display name"
+}
+
+200 OK
+{
+  "appId": "00000000-0000-0000-0000-000000000001",
+  "displayName": "a non-generated display name",
+  "foo": "testval",
+  "bar": "differentvalue",
+  ...
+}
+```
+Notes:
+1. `displayName` can be set to any value other than `null`
+2. The response body here is provided for clarity; normally on Microsoft Graph, a `204 No Content` will be the response for a `PATCH` request
```

Comment:
IMO instead of providing non-standard response it would be better to add GET request

Reply:
Hmm...

The [OData standard](https://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part1-protocol.html#sec_UpdateanEntity) says:
> Upon successful completion the service responds with either [200 OK](https://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part1-protocol.html#sec_ResponseCode200OK) and a representation of the updated entity, or [204 No Content](https://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part1-protocol.html#sec_ResponseCode204NoContent).

I don't see any guidance that actually says that we prefer `204` for `PATCH` on graph. I'm inclined to update the note to say:
> The response body here is provided for clarity, and is not part of the guidance itself. The [OData standard](https://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part1-protocol.html#sec_UpdateanEntity) states that the workload can decide the behavior. 

Does that sound alright with you?

Reply:
Sounds good! I would just replace 'OData standard' with 'Graph standard'